File: 250. Count Univalue Subtrees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def countUnivalSubtrees(self, root: TreeNode) -> int:
        if not root:
            return 0
        self.count = 0
        self.counttree(root)
        return self.count
    
    def counttree(self, root):
        if not root.left and not root.right:
            self.count += 1
            return root.val
        elif not root.left:
            if root.val == self.counttree(root.right):
                self.count += 1
                return root.val
            else:
                return
        elif not root.right:
            if root.val == self.counttree(root.left):
                self.count += 1
                return root.val
            else:
                return
        else:
            left = self.counttree(root.left)
            right = self.counttree(root.right)
            if root.val == left and root.val == right:
                self.count += 1
                return root.val
            else:
                return

File: 250. Count Univalue Subtrees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int ans;
    public int countUnivalSubtrees(TreeNode root) {
        ans=0;
        helper(root);
        return ans;
    }
    private boolean helper(TreeNode node) {
        if (node == null) {
            return true;
        }
        boolean left = helper(node.left);
        boolean right = helper(node.right);
        if (left && right) {
            if ((node.left==null || node.left.val == node.val) && (node.right==null || node.right.val==node.val)) {
                ans++;
                return true;
            } else {
                return false;
            }
        } else {
            return false;
        }
    }
}

class Solution {
    int ans;
    public int countUnivalSubtrees(TreeNode root) {
        ans=0;
        helper(root, 0);
        return ans;
    }
    private boolean helper(TreeNode node, int parVal) {
        if (node==null) {
            return true;
        }
        boolean left = helper(node.left, node.val);
        boolean right = helper(node.right, node.val);
        if (!left || !right) {
            return false;
        }
        ans++;
        return node.val == parVal;
    }
}